File: leetcode/377.cpp
class Solution {
public:
    unordered_map<int, int> dp = {{0,1}};
    int combinationSum4(vector<int>& nums, int target) {
        if(target < 0) return 0;
        if(dp.find(target) != dp.end()) return dp[target];
        int ans = 0;
        for(auto &num : nums) {
            /*
            f(nums, 4) = sum([f(nums, 4-nums[0]), f(nums, 4-nums[1]),
                            ...f(nums, 4-nums[i]))
            */
            ans += combinationSum4(nums, target-num);
        }
        dp[target] = ans;
        return ans;
    }
};

File: leetcode/377.py
'''
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<long long> cur(target+1, 0);
        cur[0] = 1;
        for(int i = 0; i < target; ++i) {
            for(auto num : nums) {
                if(i + num <= target) {
                    cur[i+num] += cur[i];
                }
            }
        }
        return cur[target];
    }
};
'''


class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        cur = [0] * (target + 1)
        cur[0] = 1
        for i in range(0, target):
            for num in nums:
                if (i + num) <= target:
                    cur[(i+num)] += cur[i]
        return cur[target]
